forest/src/main.rs

```diff
@@ -6,23 +6,118 @@ mod daemon;
+// shared memory. The identity of the shared memory object is written to a local
+// file named .forest_daemon_ipc. The parent process is responsible for cleaning
+// up the local file and the shared memory object.
+fn ipc_shmem_conf() -> ShmemConf {
+    ShmemConf::new()
+        .size(Event::size_of(None))
+        .force_create_flink()
+        .flink(std::env::temp_dir().join("forest_daemon_ipc"))
+}
+
+// Initiate an Event object in shared memory.
+fn create_ipc_lock() {
+    let mut shmem = ipc_shmem_conf().create().expect("create must succeed");
+    // The shared memory object will not be deleted when 'shmen' is dropped
+    // because we're not the owner.
+    shmem.set_owner(false);
+    unsafe {
+        Event::new(shmem.as_ptr(), true).expect("new must succeed");
+    }
+}
+
+fn build_daemon<'a>(config: &DaemonConfig) -> anyhow::Result<Daemon<'a>> {
+    let mut daemon = Daemon::new()
+        .umask(config.umask)
+        .work_dir(&config.work_dir)
+        .stdout(File::create(&config.stdout)?)
+        .stderr(File::create(&config.stderr)?);
+    if let Some(user) = &config.user {
+        daemon = daemon.user(User::try_from(user)?)
+    }
+    if let Some(group) = &config.group {
+        daemon = daemon.group(Group::try_from(group)?)
+    }
+    if let Some(path) = &config.pid_file {
+        daemon = daemon.pid_file(path, Some(false))
+    }
+
+    daemon = daemon.setup_post_fork_parent_hook(|_parent_pid, _child_pid| {
+        let mut shmem = ipc_shmem_conf().open().expect("open must succeed");
+        shmem.set_owner(true);
+        let (event, _) =
+            unsafe { Event::from_existing(shmem.as_ptr()).expect("from_existing must succeed") };
+        let ret = event.wait(EVENT_TIMEOUT);
+        drop(shmem); // Delete the local link and the shared memory object.
+        if let Err(e) = ret {
+            cli_error_and_die(
+                format!(
+                    "Error unblocking process. Error was: {e}. Check the log file for details."
+                ),
+                1,
+            );
+        }
+        info!("Forest has been detached and runs in the background.");
+        process::exit(0);
+    });
+
+    Ok(daemon)
+}
+
+fn main() {
     logger::setup_logger();
+
     // Capture Cli inputs
     let Cli { opts, cmd } = Cli::from_args();
 
     // Run forest as a daemon if no other subcommands are used. Otherwise, run the subcommand.
     match opts.to_config() {
         Ok(cfg) => match cmd {
-            Some(command) => subcommand::process(command, cfg).await,
-            None => daemon::start(cfg).await,
+            Some(command) => {
+                task::block_on(subcommand::process(command, cfg));
+            }
+            None => {
+                if opts.detach {
+                    create_ipc_lock();
+                    info!(
+                        "Redirecting stdout and stderr to files {:?} and {:?}.",
```

Comment:
how about a regular `Display` and not `Debug`? Not sure how it looks like in `Debug`.

Reply:
We already doing it at other places: Ie: `info!("Recovered libp2p keypair from {:?}", &path);` resulting in `2022-09-05T14:07:40.373Z INFO  forest_libp2p::service > Recovered libp2p keypair from "/Users/guillaume/Library/Application Support/com.ChainSafe.Forest/libp2p/keypair"`

Both `PathBuff` and `Path` don't implement `Display` trait btw.

Reply:
Ya, I know they don't implement it, but you can call e.g. `to_string_lossy`. `Debug` should be left for `debug!` :)